### PR TITLE
added `on_backward` trainer callback

### PR DESCRIPTION
**Pull request by ArjunSubramonian**
Additions proposed in this pull request:
- Added `on_backward` training callback which allows for control over backpropagation and gradient manipulation.

### CODE REVIEW

**Review by dirkgr — CHANGES_REQUESTED** [review]
I think this design is too heavy. I don't like putting a core functionality like `loss.backward()` into a callback. It makes it too hard to see what's going on in the trainer.

Instead, can we use the normal `TrainerCallback`, and give it some extra methods, like `pre_backward()` and `post_backward()`? Can you solve your problem with that?

**Comment by ArjunSubramonian** [comment]
> I think this design is too heavy. I don't like putting a core functionality like `loss.backward()` into a callback. It makes it too hard to see what's going on in the trainer.
> 
> Instead, can we use the normal `TrainerCallback`, and give it some extra methods, like `pre_backward()` and `post_backward()`? Can you solve your problem with that?

I can't do adversarial training without `backward()` being called in a callback because I need to call `backward()` two times and one call requires `retain_graph=True`.

**Comment by dirkgr** [comment]
You could leave one `backward()` call as it is, and issue the second one in the callback?

**Comment by ArjunSubramonian** [comment]
> You could leave one `backward()` call as it is, and issue the second one in the callback?

@dirkgr I have made the revisions you suggested :)

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## Unreleased
 
+### Added
+
+- Added `BackwardCallback`, a training callback which allows for control over backpropagation and gradient manipulation.
```
Comment (dirkgr) [inline CHANGELOG.md:12]: This comment isn't accurate anymore, is it?

File: allennlp/training/callbacks/backward.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+        self,
+        trainer: "GradientDescentTrainer",
+        loss: torch.FloatTensor,
+        backward_called: bool,
+        **kwargs
+    ) -> bool:
+        if not backward_called:
+            trainer._scaler.scale(loss).backward()  # type: ignore
+            return True
+        return False
```
Comment (dirkgr) [inline allennlp/training/callbacks/backward.py:26]: Is it an error if this gets called with `backward_called == True`? Should we throw an exception in that case?

File: tests/training/trainer_test.py [hunk 3]
```diff
@@ -127,6 +127,26 @@ def setup_method(self):
...
+        self,
+        trainer: "GradientDescentTrainer",
+        loss: torch.FloatTensor,
+        backward_called: bool,
+        **kwargs,
+    ) -> bool:
+        if not backward_called:
+            loss.backward()
+            for param in trainer.model.parameters():
+                param.grad *= 0.0
```
Comment (dirkgr) [inline tests/training/trainer_test.py:145]: Is that really the best way to do that?
```suggestion
                param.zero_()
```
I don't know for sure, but I would guess that `zero_()` is faster.

**Review by dirkgr — APPROVED** [review]
Other than the changelog entry, this is great!

**Review by dirkgr — APPROVED** [review]
This is great! Will make a great update for tomorrow's meeting, too!
